Recommend against EPSILON for equality tests, align related pages (#20493)
mdn/content: files/en-us/web/javascript/reference/global_objects/number/epsilon/index.md

The **`Number.EPSILON`** property represents the difference between 1 and the smallest floating point number greater than 1.

Because `EPSILON` is a static property of {{jsxref("Number")}}, you always use it as `Number.EPSILON`, rather than as a property of a number value.

{{EmbedInteractiveExample("pages/js/number-epsilon.html")}}{{js_property_attributes(0, 0, 0)}}

## Description

> The `EPSILON` property has a value of approximately `2.2204460492503130808472633361816E-16`, or 2<sup>-52</sup>. This is the smallest value that can be added to 1 to get a distinct number, because [double precision floating point format](https://en.wikipedia.org/wiki/Double_precision_floating-point_format) only has 53 bits (with the highest bit always being 1) to represent the significand, and the lowest bit has a significance of 2<sup>-52</sup>.

Reviewer: do people know what significand means? Is it worth linking to a definition or explaining?

Author: True... I don't know a better way to describe "significand". What about:

```suggestion
The `EPSILON` property has a value of approximately `2.2204460492503130808472633361816E-16`, or 2<sup>-52</sup>. This is the smallest value that can be added to 1 to get a distinct number, because [double precision floating point format](https://en.wikipedia.org/wiki/Double_precision_floating-point_format) only has 53 bits (with the highest bit always being 1) to represent the significant digits (significand), and the lowest bit has a significance of 2<sup>-52</sup>.
```

I think "significant digits" has a wider understanding.

Note that the absolute accuracy of floating numbers decreases as the number gets larger, because the exponent grows while the significand's accuracy stays the same. {{jsxref("Number.MIN_VALUE")}} is the smallest representable positive number, which is much smaller than `Number.EPSILON`.

## Examples

### Testing equality

Because floating point numbers are represented in binary, not decimal, they are not always exact. For example, `0.1 + 0.2` is not exactly equal to `0.3`:

```js
console.log(0.1 + 0.2); // 0.30000000000000004
console.log(0.1 + 0.2 === 0.3); // false
```

> For this cause, it is often advised that **floating point numbers should never be compared with `===`**. Instead, we can deem two numbers as equal if they are _close enough_ to each other. The `Number.EPSILON` constant is usually a reasonable threshold for errors if the arithmetic is around the magnitude of `1`, because `EPSILON`, in essence, specifies how accurate the number "1" is.

Reviewer: ```suggestion
For this reason, it is often advised that **floating point numbers should never be compared with `===`**. Instead, we can deem two numbers as equal if they are _close enough_ to each other. The `Number.EPSILON` constant is usually a reasonable threshold for errors if the arithmetic is around the magnitude of `1`, because `EPSILON`, in essence, specifies how accurate the number "1" is.
```

```js
function equal(x, y) {
  return Math.abs(x - y) < Number.EPSILON;
}

const x = 0.2;
const y = 0.3;
const z = 0.1;
console.log(equal(x + z, y)); // true
```

> However, `Number.EPSILON` is inappropriate for any arithmetic with a larger range. If your data is on the 10<sup>3</sup> order of magnitude, the decimal part will have a much smaller accuracy than `Number.EPSILON`:

Reviewer: ```suggestion
However, `Number.EPSILON` is inappropriate for equations with a larger range. If your data is on the 10<sup>3</sup> order of magnitude, the decimal part will have a much smaller accuracy than `Number.EPSILON`:
```

Reviewer: not a fan of the word "range" because I don't know exactly what you mean, and since I don't know what you mean, I don't have a better alternative.

Author: Let's say "magnitude". Also I think "arithmetic" is better than "equation", because "equation" is an entity (like "expression") while "arithmetic" sounds more like an action.

```suggestion
However, `Number.EPSILON` is inappropriate for any arithmetic operating on a larger magnitude. If your data is on the 10<sup>3</sup> order of magnitude, the decimal part will have a much smaller accuracy than `Number.EPSILON`:
```

```js
function equal(x, y) {
  return Math.abs(x - y) < Number.EPSILON;
}

const x = 1000.1;
const y = 1000.2;
const z = 2000.3;
console.log(x + y); // 2000.3000000000002; error of 10^-13 instead of 10^-16
console.log(equal(x + y, z)); // false
```

> In this case, you should use a larger tolerance, such as `2000 * Number.EPSILON`, since the numbers we are comparing have the magnitude of about `2000`:

Reviewer: ```suggestion
In such a case, a larger tolerance is required. And the numbers compared have a magnitude of approximately `2000`, a multiplier, such as `2000 * Number.EPSILON`, creates a robust enough tolerance.  
```

Author: ```suggestion
In this case, a larger tolerance is required. As the numbers compared have a magnitude of approximately `2000`, a multiplier such as `2000 * Number.EPSILON` creates enough tolerance.
```

(I don't want to say it's "robust" because the number getting any larger than that will make it fail again...)

Reviewer: how about `creates enough tolerance for this instance.`?

Author: Sounds good👍

```suggestion
In this case, a larger tolerance is required. As the numbers compared have a magnitude of approximately `2000`, a multiplier such as `2000 * Number.EPSILON` creates enough tolerance for this instance.
```

```js
function equal(x, y, tolerance = Number.EPSILON) {
  return Math.abs(x - y) < tolerance;
}

const x = 1000.1;
const y = 1000.2;
const z = 2000.3;
console.log(equal(x + y, z, 2000 * Number.EPSILON)); // true
```

> Besides magnitude, it's important to also consider the _accuracy_ of your input. For example, if the numbers are collected from a form input, and the input value can only be adjusted by steps of `0.1`, it usually makes sense to allow a much larger tolerance, such as `0.01`, since the data only has a precision of `0.1`.

Reviewer: ```suggestion
In addition to magnitude, it is important to consider the _accuracy_ of your input. For example, if the numbers are collected from a form input and the input value can only be adjusted by steps of `0.1` (i.e. `[<input type="number" step="0.1">](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/step)`), it usually makes sense to allow a much larger tolerance, such as `0.01`, since the data only has a precision of `0.1`.
```

> > **Note:** Important takeaway: do not blindly use `Number.EPSILON` as a threshold for equality testing. Use a threshold that is appropriate for the magnitude and accuracy of the numbers you are comparing.

Reviewer: ```suggestion
> **Note:** Important takeaway: do not simply use `Number.EPSILON` as a threshold for equality testing. Use a threshold that is appropriate for the magnitude and accuracy of the numbers you are comparing.
```

Reviewer: not a fan of either term.... 

maybe 

`Important takeaway: Always consider `Number.EPSILON`
s limitations as a threshold for equality testing`
